### PR TITLE
Unify_certs: New certificate generation scheme with verification

### PR DESCRIPTION
**Pull request by situ-s**
This PR handles:

1. Replica & Bft Client new certificate generation scheme
2. Code modification to verify the new scheme
3. Directory structure changed from <Replica_id>/<node_type>/<node_type>.cert where node_type can be "client" or "server" to <Replica_id>/tls.cert
4. Flag useUnifiedCertificates added for unification of certificates.

### CODE REVIEW

File: communication/CMakeLists.txt [hunk 1]
```diff
@@ -22,11 +22,13 @@ add_library(bftcommunication ${bftcommunication_src})
 add_library(bftcommunication_shared SHARED ${bftcommunication_src})
 set_property(TARGET bftcommunication_shared PROPERTY POSITION_INDEPENDENT_CODE ON)
 
+target_include_directories(bftcommunication PUBLIC ../bftengine/include/bftengine)
```
Comment (yuliasherman) [inline communication/CMakeLists.txt:25]: Please don't use relative directory paths in CMakeLists files.

File: communication/src/AsyncTlsConnection.cpp [hunk 2]
```diff
@@ -283,8 +287,10 @@ void AsyncTlsConnection::initClientSSLContext(NodeNum destination) {
   ssl_context_.set_verify_mode(asio::ssl::verify_peer);
 
   fs::path path;
+  fs::path cert_path;
   try {
-    path = fs::path(config_.certificatesRootPath_) / fs::path(std::to_string(config_.selfId_)) / "client";
+    path = fs::path(config_.certificatesRootPath_) / fs::path(std::to_string(config_.selfId_));
+    if (bftEngine::ReplicaConfig::instance().enableMultiplexChannel == false) path = path / fs::path("client");
```
Comment (yuliasherman) [inline communication/src/AsyncTlsConnection.cpp:293]: Condition in form 
if (!bftEngine::ReplicaConfig::instance().enableMultiplexChannel)
is preferable.

Comment (situ-s) [reply communication/src/AsyncTlsConnection.cpp]: done

File: communication/src/AsyncTlsConnection.cpp [hunk 3]
```diff
@@ -27,8 +27,12 @@
 #include "crypto_utils.hpp"
 #include "communication/StateControl.hpp"
 #include "hex_tools.h"
+#include "ReplicaConfig.hpp"
+
 namespace bft::communication::tls {
 
+using bftEngine::ReplicaConfig;
```
Comment (yuliasherman) [inline communication/src/AsyncTlsConnection.cpp:34]: No need in this

File: communication/src/AsyncTlsConnection.cpp [hunk 4]
```diff
@@ -283,8 +287,10 @@ void AsyncTlsConnection::initClientSSLContext(NodeNum destination) {
   ssl_context_.set_verify_mode(asio::ssl::verify_peer);
 
   fs::path path;
+  fs::path cert_path;
   try {
-    path = fs::path(config_.certificatesRootPath_) / fs::path(std::to_string(config_.selfId_)) / "client";
+    path = fs::path(config_.certificatesRootPath_) / fs::path(std::to_string(config_.selfId_));
+    if (!bftEngine::ReplicaConfig::instance().useUnifiedCertificates) path = path / fs::path("client");
```
Comment (yuliasherman) [inline communication/src/AsyncTlsConnection.cpp:293]: Define a static data member variable for this in AsyncTlsConnection class and set it in static std::shared_ptr<AsyncTlsConnection> create() function.

File: communication/src/AsyncTlsConnection.cpp (hunk 3 above)
Comment (situ-s) [reply communication/src/AsyncTlsConnection.cpp]: oh my mistake

File: communication/src/AsyncTlsConnection.cpp (hunk 4 above)
Comment (situ-s) [reply communication/src/AsyncTlsConnection.cpp]: made the changes.

**Review by yuliasherman — APPROVED** [review]
LGTM

File: communication/src/AsyncTlsConnection.cpp [hunk 5]
```diff
@@ -27,8 +27,11 @@
 #include "crypto_utils.hpp"
 #include "communication/StateControl.hpp"
 #include "hex_tools.h"
+
 namespace bft::communication::tls {
 
+bool AsyncTlsConnection::useUnifiedCertificates_;
```
[the remark this answers is not in the document]
Comment (situ-s) [reply communication/src/AsyncTlsConnection.cpp]: Thanks for pointing this.

**Pull request merged**